### DISCUSSION TITLE
Ability to rename an environment

### DISCUSSION THREAD

**Discussion by greggbjensen** [community/community#64081] [Actions]
### Select Topic Area

Product Feedback

### Body

We have many secrets set up in our environments, and unfortunately, the initial ones we created start with an uppercase letter.  Now we have standardized on all lowercase letters to make common actions easier.  However, we cannot rename an existing environment, clone it, or copy the secrets.

It would be great if we could rename an environment to fix this.

**Comment by christophercouch** [comment]
Hi @greggbjensen,

I see the challenge you're facing. Having the ability to rename an environment, especially when you're standardizing naming conventions, would indeed be a helpful feature. This would also simplify the process of managing secrets and environments without having to recreate everything from scratch.

I'd recommend submitting this as a feature request directly through the [GitHub feedback form](https://support.github.com/contact/feedback). This ensures the GitHub team can evaluate your suggestion.

Thanks for sharing your feedback!

**Reply by risoms** [reply]
@christophercouch Has there been any update to this? I can imagine there is a shared need of this across many users.

**Reply by amigos-sian** [reply]
hi @christophercouch , also keen for an update on renaming environments. Reason for this is that we have environments that we no longer needed, but you wouldn't want to delete them as the environments provide a record of past deployments which is important for audits etc. To further define the issue: "As a developer, given I have clicked on an environment, under Settings/Environments for a repo, I would to click on an option to rename an environment under the configuration settings".

**Reply by AnijaJacob** [reply]
+1 for this enhancement request.
Given the flexibility with GHA , we are using same as the single orchestrator for all deployments . Manuel migration to new environments with 40+ vars instead of a simple name change would be time consuming

**Reply by leymbda** [reply]
+1 as well. I don't have as many vars and secrets but would still be a pain!

**Reply by cmyers009** [reply]
+1 I would also like this.  We set up our environment and want to rename it.

**Reply by thunermay** [reply]
+3 (3 developers): initialized as "production" and have to rename it for the sake of terraform to "prod"

**Reply by vuolo** [reply]
+2

**Reply by ilchenkoArtem** [reply]
+1

**Reply by greggbjensen** [reply]
Thanks for the reply!

**Reply by Arno-ATU** [reply]
+1 any updates on this?

**Reply by jespernohr** [reply]
+1

**Reply by nickgrealy** [reply]
+1

**Comment by nebuk89** [comment]
This is a good idea, I will add it to our paper cuts list to try to get to. We are getting pretty full for this year (we have a bunch of paper cuts :D ) but I will ensure this is in the list and considered <3

**Reply by nebuk89** [reply]
I would <3 y'alls thoughts on how this should interact with the fact that if a workflow references an environment that doesn't exist. Github Actions creates it. I guess my worry here is that files being missed in rename would just 'create' new environments.

Or should we just block this behavior 👀 
 https://github.com/orgs/community/discussions/84203

**Reply by cmyers009** [reply]
> I would <3 y'alls thoughts on how this should interact with the fact that if a workflow references an environment that doesn't exist. Github Actions creates it. I guess my worry here is that files being missed in rename would just 'create' new environments.
> 
> Or should we just block this behavior 👀 https://github.com/orgs/community/discussions/84203

I am thinking that if the environment does not exist and is being referenced, it should throw either a warning or an error, because I am not sure what the point of creating a new, empty environment would be.

it seems like the behavior as you describe makes the potential issue (incorrectly named environments) harder to find than if it failed in build step. 


This next part may be out of scope, but I think it would be a really cool idea if GitHub could detect recently changed environment names, and then give a warning on the screen like “You recently changed the environment name for previous_environment to new_envirnment. You are still referencing the old environment name.  {list of links to take you to that location in your workflow file} 
with an option to dismiss so you don’t get notified of the issue again.

**Reply by shariquealavi** [reply]
The automatic creation is an unexpected behavior anyway ... that too with non existent or wrong values ... so it's better to error out rather than auto create.

**Comment by iCount** [comment]
This is so good idea

**Comment by donnaken15** [comment]
I'm wondering about this too because I'm trying to test actions and deployments are getting labelled as the environments I made, which I named without thinking they would be used for that

**Comment by t9luke** [comment]
Apologies for the bump, but are there any updates? I have one environment where the name no longer makes sense for its current use. I’d recreate it, but since I don’t have access to the original secrets, being able to rename it would be a huge help.

**Comment by tonytromp02** [comment]
Yes, also wanted this to be as simple as edit name. But seems AI features have become more important than basic requests :(

**Comment by aguerrerott** [comment]
Still no updates 💔🥀

**Comment by h-lakshman** [comment]
Are there any updates on this?

**Comment by wasifali758595** [comment]
The ability to rename an environment depends on the platform you’re working in. In Salesforce (or similar systems), once an environment (such as a sandbox, production, or developer org) is created, its system name cannot be changed. However, you can:

Rename the display label – For example, in sandboxes you can update the Sandbox Name and Description from Setup > Sandboxes, which helps differentiate environments.

Use custom naming conventions – Many teams use consistent prefixes/suffixes (e.g., UAT_Test, Dev_Jane) to make environments easy to recognize.

Create a new environment – If you truly need a different environment name at the system level, the only option is to spin up a new one with the desired name.

In short: You cannot rename an environment itself once it’s created, but you can rename its label/description for clarity.

**Reply by h-lakshman** [reply]
If GitHub uses secret names as unique identifiers, I'm curious how Bitbucket's similar repository variables can be renamed. A developer might have effectively lost the key to a secret, and without the ability to rename it, the only alternative deleting and recreating the secret is not possible. For this reason, a rename feature would have been a valuable convenience.

**Comment by federico-balice** [comment]
Yet another update ask here  ¯\_ (ツ)_/¯
Changing the environment's names is pretty common in the industry, so it's kinda strange that this has not been implemented yet.
Maybe there are good reasons not to have this in, but in that case I'd love to hear the reason.

Looking forward for an update.
Kindest regards,
    Fede

**Comment by aguerrerott** [comment]
still no updates again 💔🥀

**Reply by AbdelRahmanRahal** [reply]
Yup. Thought I'd scroll down to find it fixed but... yeah.

**Comment by nsuman** [comment]
Any updates on this ?

**Reply by donnaken15** [reply]
of course not

only more AI being pushed...

**Comment by Daniel217D** [comment]
2026!

**Reply by rzavadzich** [reply]
Please

**Comment by okohll** [comment]
A vote for this

**Comment by spereira-bt** [comment]
+1 vote

**Comment by mmaryo** [comment]
+1

**Comment by Shahwaiz-Dev** [comment]
+1000

**Comment by stupidly-logical** [comment]
+1001

**Comment by shariquealavi** [comment]
Is this really something that'd take more than 3 years?

**Reply by Mr-Carrot0** [reply]
Friendly reminder that Github is owned by Microslop

**Comment by theking2** [comment]
What is the road map for this feature? It seems like this can be done by copilot in 15-30sec.

**Comment by aabulimiti** [comment]
+1

**Comment by pasanchamikara** [comment]
Bumping this!

**Comment by kiwifellows** [comment]
And sorry to say it is 2026 and still not implemented. Looking forward to this as well
:+1:

**Comment by marinamedeirosz** [comment]
+1

**Comment by kiwifellows** [comment]
I found out something very interesting... environment names are not case sensitive and so workaround is straightforward, but needs to be documented as such. If someone points me to some docs, I can do a PR.

Example from my GH Action:
```YAML
workflow_dispatch:
    inputs:
      environment:
        description: Target GitHub Environment (dev / uat / prod)
        required: true
        type: choice
        options: [dev, uat, prod]
```

and further down:
```YAML
deploy:
    name: Deploy Cloud Run + Terraform
    needs: build-test
    if: github.event_name == 'push' || github.event_name == 'workflow_dispatch'
    runs-on: ubuntu-latest
    environment: ${{ github.event_name == 'workflow_dispatch' && inputs.environment || github.ref_name }}
    permissions:
      contents: read
    env:
      DEPLOYMENT_ENVIRONMENT: ${{ github.event_name == 'workflow_dispatch' && inputs.environment || github.ref_name }}
```
And my environment names below.
<img width="475" height="343" alt="image" src="https://github.com/user-attachments/assets/47f2ba22-22ad-4e88-91ec-a2969f0953bf" />
Have tested this works and it is perfect, so I think apart from it looking inconsistent in GH Settings UI, the main thing is at least in my GH Actions files we can use lowercase and have consistency.

**Reply by kiwifellows** [reply]
I see point 5 here does mention that environment names are NOT case sensitive so I think from an issue standpoint can this be closed? And also a link to docs be put here at the top?
https://docs.github.com/en/actions/how-tos/deploy/configure-and-manage-deployments/manage-environments?versionId=free-pro-team%40latest&productId=actions#creating-an-environment

**Comment by marijn** [comment]
I'm so happy that we live in an age where allegedly software production is so easy it has become a commodity, and that because of that basic useful features that have been open for 3 years are now dealt with at blazing speeds, creating more value than ever before...

What a time to be alive 😞

**Comment by kramber1024** [comment]
+1

**Comment by rehrl-johannes** [comment]
+1
